### ISSUE TITLE
Descriptor loss modifications question.

### ISSUE THREAD

**Issue by AliYoussef97** [rpautrat/SuperPoint#302]
Hello,

I had 2 brief questions regarding the modifications done to the descriptor loss compared to the paper, apologies if this might be a repetitive question.

I assume the descriptor loss aligns with the paper expect for:

1. The descriptors, warped_descriptors, ReLU and the descriptor dot normalizaiton. Thus, by commenting the lines continating arrows below, then the descriptor loss is aligns with the paper. As in the paper, the loss seems to be computed on D and D' which is the descriptor's head output before interpolation and normalization.

``` 
descriptors = tf.reshape(descriptors, [batch_size, Hc, Wc, 1, 1, -1])

descriptors = tf.nn.l2_normalize(descriptors, -1) <------------

warped_descriptors = tf.reshape(warped_descriptors,
                                [batch_size, 1, 1, Hc, Wc, -1])

warped_descriptors = tf.nn.l2_normalize(warped_descriptors, -1) <------------

dot_product_desc = tf.reduce_sum(descriptors * warped_descriptors, -1)

dot_product_desc = tf.nn.relu(dot_product_desc) <-----------------

dot_product_desc = tf.reshape(tf.nn.l2_normalize(
    tf.reshape(dot_product_desc, [batch_size, Hc, Wc, Hc * Wc]), <-----------------
    3), [batch_size, Hc, Wc, Hc, Wc])

dot_product_desc = tf.reshape(tf.nn.l2_normalize(
    tf.reshape(dot_product_desc, [batch_size, Hc * Wc, Hc, Wc]), <-----------------
    1), [batch_size, Hc, Wc, Hc, Wc])
```

2. The homogrophy induced "s" is was chosen as 7.5 instead of 8.

I assume these are the notable modifications done regarding the descriptors loss, unless there are other modifications that I might have missed.

Thank you for the amazing work!

**Comment by rpautrat** [comment]
Hi,

Yes I think this is correct. In addition, note that we changed two more parameters of the paper: lambda balancing the detector and descriptor loss, and lambda_d balancing positive and negative terms.

**Comment by AliYoussef97** [comment]
@rpautrat Thank you so much!

**Comment by AliYoussef97** [comment]
@rpautrat Sorry, I had another question regarding the descriptor loss.

There is something I don't quite understand. 

If the descriptor loss is a double sum of lambda_d*s*max(0,mp-dd') + (1-s)*max(0,d'd-mn) then divided by (Hc,Wc)^2.

Should not this line 

```
loss = tf.reduce_sum(valid_mask * loss) / normalization
```

be 

```
loss =  te.reduce_sum( valid_mask*(tf.reduce_sum(valid_mask * loss) )  / normalization
```

I might be completely wrong though, so apologies for the confusion.

Thank you!

**Comment by rpautrat** [comment]
Hi,

From what I remember, only the second image has bordering artifacts, so there is no need to mask the keypoints of the first image. That's why valid_mask is only of shape [batch_size, 1, 1, Hc, Wc], to mask the keypoints of the second image only, and there is only one multiplication to valid_mask. Then the tf.reduce_sum sums all dimensions, which gives you the double sum (and also a sum on the batch size as well).

**Comment by AliYoussef97** [comment]
@rpautrat Thank you so much, I think I get it :D

From what I understood, after lambda_d*s*max(0,mp-dd') + (1-s)*max(0,d'd-mn) has a size of [B,Hc,Wc,Hc,Wc], the validation mask is just to remove the bordering artifacts nothing else.

But the tf.reduce_sum(valid_mask,desc_loss),  is the double sum itself whether there the valid_mask is included or not, since the the tf.reduce_sum would sum the desc_loss over Hc,Wc,Hc,Wc which is essentially the double sum. Correct?

I think my confusion came as I thought the valid_mask had another use case other than removing the bordering artifacts, but I failed to grasp that it is just a binary mask so it is not affecting the loss in any numerical way other than removing the artifacts.

Thank you again, that was very helpful!

**Comment by rpautrat** [comment]
Yes exactly, I think you got it now :)
